Corona-Warn-App/src/main/java/de/rki/coronawarnapp/coronatest/TestCertificateRepository.kt

```diff
@@ -0,0 +1,329 @@
+                throw it
+            }
+            .launchIn(appScope + dispatcherProvider.Default)
+    }
+
+    suspend fun requestCertificate(test: CoronaTest): TestCertificateContainer {
+        Timber.tag(TAG).d("createDccForTest(test.identifier=%s)", test.identifier)
+
+        val newData = internalData.updateBlocking {
+            if (values.any { it.registrationToken == test.registrationToken }) {
+                Timber.tag(TAG).e("Certificate entry already exists for %s", test.identifier)
+                throw IllegalArgumentException("A certificate was already created for this ${test.identifier}")
+            }
+
+            // TODO do we need additional validation here? last chance to abort?
+
+            val identifier = UUID.randomUUID().toString()
+
+            val certificate = when (test.type) {
+                CoronaTest.Type.PCR -> PCRCertificateContainer(
+                    identifier = identifier,
+                    registeredAt = test.registeredAt,
+                    registrationToken = test.registrationToken,
+                )
+                CoronaTest.Type.RAPID_ANTIGEN -> RACertificateContainer(
+                    identifier = identifier,
+                    registeredAt = test.registeredAt,
+                    registrationToken = test.registrationToken,
+                )
+            }
+            Timber.tag(TAG).d("Adding test certificate entry: %s", certificate)
+            mutate { this[certificate.identifier] = certificate }
+        }
+
+        return newData.values.single { it.registrationToken == test.registrationToken }
+    }
+
+    data class RefreshResult(
+        val certificate: TestCertificateContainer,
+        val error: Exception? = null,
+    )
+
+    suspend fun refresh(identifier: TestCertificateIdentifier? = null): Set<RefreshResult> {
+        Timber.tag(TAG).d("refresh(identifier=%s)", identifier)
+
+        val refreshResults = mutableSetOf<RefreshResult>()
+
+        val workedOnIds = mutableSetOf<TestCertificateIdentifier>()
+
+        internalData.updateBlocking {
+            val toRefresh = values
+                .filter { it.identifier == identifier || identifier == null }
+                .filter { it.isPending }
+
+            mutate {
+                toRefresh.forEach {
+                    workedOnIds.add(it.identifier)
+                    this[it.identifier] = when (it.type) {
+                        CoronaTest.Type.PCR -> (it as PCRCertificateContainer).copy(isUpdatingData = true)
+                        CoronaTest.Type.RAPID_ANTIGEN -> (it as RACertificateContainer).copy(isUpdatingData = true)
+                    }
+                }
+            }
+        }
+
+        internalData.updateBlocking {
+            Timber.tag(TAG).d("Checking for unregistered public keys.")
+
+            val refreshedCerts = values
+                .filter { workedOnIds.contains(it.identifier) }
+                .filter { !it.isPublicKeyRegistered }
```

Comment:
two filters -> filter ?

Reply:
Similar to the previous double filter, i think it serves readability two have it split.
We first filter for those that are targeted by this `refresh` call, then we filter for those that we need to work on in this step.